pkg/provisioner/ironic/bios.go

```diff
@@ -0,0 +1,171 @@
+package ironic
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+)
+
+// BuildBIOSCleanSteps build clean steps for different driver
+func BuildBIOSCleanSteps(driver string, firmware *metal3v1alpha1.FirmwareConfig) (cleanSteps []nodes.CleanStep, err error) {
+	// Deal possible panic
+	defer func() {
+		r := recover()
+		if r != nil {
+			err = fmt.Errorf("panic in build BIOS settings: %v", r)
+		}
+	}()
+
+	switch driver {
+	case "ilo", "ilo5":
+		cleanSteps = buildILOCleanSteps(firmware)
+	case "irmc":
+		cleanSteps = buildIRMCCleanSteps(firmware)
+	default:
+		if firmware != nil {
+			return nil, fmt.Errorf("Firmware settings are defined, but the node's driver %s does not support configure firmware setting", driver)
+		}
+	}
+	return
+}
+
+// A private method to build the clean steps for ILO configuration from BaremetalHost spec
+func buildILOCleanSteps(firmware *metal3v1alpha1.FirmwareConfig) (cleanSteps []nodes.CleanStep) {
+	// This cleaning step resets all BIOS settings to factory default for a given node
+	cleanSteps = append(
+		cleanSteps,
+		nodes.CleanStep{
+			Interface: "bios",
+			Step:      "factory_reset",
```

Comment:
Can we make this optional? I can imagine someone only wanting to modify certain fields without a full reset.

Reply:
Thanks, done.